odoo/tests/common.py

```diff
@@ -415,18 +415,20 @@ def with_user(self, login):
     @contextmanager
     def _assertRaises(self, exception, *, msg=None):
         """ Context manager that clears the environment upon failure. """
-        with super(BaseCase, self).assertRaises(exception, msg=msg) as cm:
+        with super().assertRaises(exception, msg=msg) as cm, ExitStack() as ex:
             if hasattr(self, 'env'):
-                with self.env.cr.savepoint():
+                try:
+                    ex.enter_context(self.env.cr.savepoint())
                     if issubclass(exception, AccessError):
                         # The savepoint() above calls flush(), which leaves the
                         # record cache with lots of data.  This can prevent
                         # access errors to be detected. In order to avoid this
                         # issue, we clear the cache before proceeding.
                         self.env.cr.clear()
-                    yield cm
-            else:
-                yield cm
+                except exception:
```

Comment:
I suggest using another variable name than `exception` here.

Reply:
It's how the method parameter is called, didn't see a need to rename it :shrug:

Reply:
Ouch, this confused me: I thought that `exception` was the raised exception instead of the caught exception type!
So that's correct indeed.